File: adk-examples/Python/adk-connector.py
from falkonryclient import client as Falkonry
from falkonryclient import schemas as Schemas
from fileAdapter import FileAdapter
from multiprocessing import Process
import io
import os
import json
import time as timepkg
import logging


logging.basicConfig(
    level=logging.DEBUG,
    format="%(asctime)s:%(levelname)s:%(message)s"
    )

class AddDataException(Exception):
    pass


class ADKconn:
    url = "https://dev.falkonry.ai"
    token = "7gjg7pjryg6mry2ypq6jbhgd9kg2rn2q"
    falkonry   = Falkonry(url, token)
    datastream = Schemas.Datastream()
    datasource = Schemas.Datasource()
    field = Schemas.Field()
    time = Schemas.Time()
    signal = Schemas.Signal()
    datastreamId = None


    def checkDataIngestion(self, tracker):
        tracker_obj = None
        for i in range(0, 12):
            tracker_obj = self.falkonry.get_status(tracker['__$id'])
            logging.info(tracker_obj['status'])
            try:
                if tracker_obj['status'] == 'FAILED' or tracker_obj['status'] == 'ERROR':
                    raise Exception()
                if tracker_obj['status'] == 'COMPLETED' or tracker_obj['status'] == 'SUCCESS':
                    logging.info("Data added successfully.")
                    return "SUCCESS"
            except:
                pass
            timepkg.sleep(5)

        if tracker_obj['status'] == 'PENDING':
            raise AddDataException()


    def postMoreDataFromStream(self, datastreamId, path, liveStatus):
        onlyfiles = [f for f in os.listdir(path) if ((os.path.isfile(os.path.join(path, f)) and ((os.path.splitext(f))[1] == ".csv") or (os.path.splitext(f))[1] == ".json"))]
        onlyfiles_length = len(onlyfiles)
        for i in range(onlyfiles_length):
            datastreamId = datastreamId
            file_adapter = FileAdapter()
            data, fileType = file_adapter.getData(path + "/" + onlyfiles[i])
            datastream = self.falkonry.get_datastream(datastreamId)
            options = {
                'streaming': liveStatus,
                'hasMoredata': True,
                'timeFormat': datastream.get_field().get_time().get_format(),
                'timeZone': datastream.get_field().get_time().get_zone(),
                'timeIdentifier': datastream.get_field().get_time().get_identifier(),
                # 'signalIdentifier': datastream.get_field().get_signal().get_signalIdentifier(),  ### Narrow Datastream Format
                # 'valueIdentifier': datastream.get_field().get_signal().get_valueIdentifier(),    ### Narrow Datastream Format
                # 'batchIdentifier': datastream.get_field().get_batchIdentifier(),                 ### Batch window type
                'entityIdentifier': datastream.get_field().get_entityIdentifier()
            }

            if i == onlyfiles_length - 1:
                options["hasMoreData"] = False

            if not liveStatus:
                for i in range(0, 3):
                    try:
                        inputResponse = self.falkonry.add_input_data(datastreamId, fileType, options, data)
                        status = self.checkDataIngestion(inputResponse)
                        if status == "SUCCESS":
                            break
                    except AddDataException:
                        logging.warning("Adding data failed! Retrying({})".format(i + 1))
                if i == 3:
                    raise Exception("Cannot add data to the datastream!")
            else:
                inputResponse = self.falkonry.add_input_data(datastreamId, fileType, options, data)

    def createDataStream(self):
        name = "SL_MIXED_MUL_PY_1232763"
        self.datastream.set_name(name)

        timezone = 'GMT'
        self.time.set_zone(timezone)

        timeIdentifier = "time"
        self.time.set_identifier(timeIdentifier)

        self.time.set_format("millis")

        precisionFormat = 'millis'
        self.datastream.set_time_precision(precisionFormat)

        self.field.set_time(self.time)
        self.field.set_signal(self.signal)
        self.datasource.set_type("STANDALONE")
        self.datastream.set_datasource(self.datasource)
        self.datastream.set_field(self.field)

        entityName = "entity"
        self.field.set_entityIdentifier(entityName)

        ##############  For narrow datastream format ################
        # signalIdentifier = "signal"
        # valueIdentifier = "value"
        # self.signal.set_signalIdentifier(signalIdentifier)
        # self.signal.set_valueIdentifier(valueIdentifier)

        ################  For batch window Type  ###################
        # batchIdentifier = "batch_id"
        # self.field.set_batchIdentifier(batchIdentifier)

        createdDataStream = self.falkonry.create_datastream(self.datastream)
        datastreamId = createdDataStream.get_id()
        return datastreamId


    def postData(self, datastreamId, data, fileType, liveStatus):
        datastreamId = datastreamId
        datastream = self.falkonry.get_datastream(datastreamId)

        options = {
        'streaming': liveStatus,
        'hasMoreData': False,
        'timeFormat': datastream.get_field().get_time().get_format(),
        'timeZone': datastream.get_field().get_time().get_zone(),
        'timeIdentifier': datastream.get_field().get_time().get_identifier(),
        # 'signalIdentifier': datastream.get_field().get_signal().get_signalIdentifier(),  ### Narrow Datastream Format
        # 'valueIdentifier': datastream.get_field().get_signal().get_valueIdentifier(),    ### Narrow Datastream Format
        # 'batchIdentifier': datastream.get_field().get_batchIdentifier(),                 ### Batch window type
        'entityIdentifier': datastream.get_field().get_entityIdentifier()
        }

        if not liveStatus:
            for i in range(0, 3):
                try:
                    inputResponse = self.falkonry.add_input_data(datastreamId, fileType, options, data)
                    status = self.checkDataIngestion(inputResponse)
                    if status == "SUCCESS":
                        break
                except AddDataException:
                    logging.warning("Adding data failed! Retrying({})".format(i + 1))
            if i == 3:
                raise Exception("Cannot add data to the datastream!")
        else:
            inputResponse = self.falkonry.add_input_data(datastreamId, fileType, options, data)


    def postDataFromStream(self, datastreamId, data, fileType, liveStatus):
        file_name = "simple_sliding_mixed_multi_entity_source_moreData.csv"
        datastreamId = datastreamId
        datastream = self.falkonry.get_datastream(datastreamId)
        options = {
            'streaming': liveStatus,
            'hasMoreData': False,
            'timeFormat': datastream.get_field().get_time().get_format(),
            'timeZone': datastream.get_field().get_time().get_zone(),
            'timeIdentifier': datastream.get_field().get_time().get_identifier(),
            # 'signalIdentifier': datastream.get_field().get_signal().get_signalIdentifier(),  ### Narrow Datastream Format
            # 'valueIdentifier': datastream.get_field().get_signal().get_valueIdentifier(),    ### Narrow Datastream Format
            # 'batchIdentifier': datastream.get_field().get_batchIdentifier(),                 ### Batch window type
            'entityIdentifier': datastream.get_field().get_entityIdentifier()
        }

        if not liveStatus:
            for i in range(0, 3):
                try:
                    inputResponse = self.falkonry.add_input_data(datastreamId, fileType, options, data)
                    status = self.checkDataIngestion(inputResponse)
                    if status == "SUCCESS":
                        break
                except AddDataException:
                    logging.warning("Adding data failed! Retrying({})".format(i + 1))
            if i == 3:
                raise Exception("Cannot add data to the datastream!")
        else:
            inputResponse = self.falkonry.add_input_data(datastreamId, fileType, options, data)


    def postRealtimeData(self, datastreamId, data, fileType):
        file_name = "simple_sliding_mixed_multi_entity_source_moreData.csv"
        datastreamId = datastreamId

        datastream = self.falkonry.get_datastream(datastreamId)
        options = {
            'streaming': True,
            'hasMoreData': False,
            'timeFormat': datastream.get_field().get_time().get_format(),
            'timeZone': datastream.get_field().get_time().get_zone(),
            'timeIdentifier': datastream.get_field().get_time().get_identifier(),
            # 'signalIdentifier': datastream.get_field().get_signal().get_signalIdentifier(),  ### Narrow Datastream Format
            # 'valueIdentifier': datastream.get_field().get_signal().get_valueIdentifier(),    ### Narrow Datastream Format
            # 'batchIdentifier': datastream.get_field().get_batchIdentifier(),                 ### Batch window type
            'entityIdentifier': datastream.get_field().get_entityIdentifier()
        }

        # for i in range(5):
        #     fileName, fileExtension = os.path.splitext(file_name)
        #     stream = io.open(fileName + str(i) + fileExtension)
        #     inputResponse = self.falkonry.add_input_stream(datastreamId, fileType, options, stream)
        #     df = pd.read_csv(fileName + str(i) + fileExtension)
        #     time_colm = df.loc[:, "time"]
        #     prev_time = df.loc[len(df) - 1, "time"]
        #     time_difference = df.loc[1, "time"] - df.loc[0, "time"]
        #     for j in range(len(time_colm)):
        #         if j == 0:
        #             df.loc[j, "time"] = prev_time + time_difference
        #         else:
        #             df.loc[j, "time"] = df.loc[j - 1, "time"] + time_difference
        #
        #     df = df.loc[:, ]
        #
        #     df.to_csv(fileName + str(i + 1) + fileExtension, index=False)

        
        inputResponse = self.falkonry.add_input_stream(datastreamId, fileType, options, data)



    def getLiveOutput(self, assessmentId):
        assessmentId = assessmentId
        stream = self.falkonry.get_output(assessmentId, None)
        for event in stream.events():
            logging.info(json.dumps(json.loads(event.data)))



if __name__ == "__main__":

    # File adapter will be used to get the appropriate data from the different files and to provide it
    # in the format in which we can send to the ADK methods.


    #################### For creating datastream and adding historical data ################
    """
    The below code will create a datastream and post the historical data as a string to the
    datastream. You will have to give the data as a string or give the fileName and pass it 
    to the get_data() method of the file adapter.
    """

    # fileAdapter = FileAdapter()
    # fileName = "SL_MIXED_MUL_PY_12345.json"
    # data, fileType = fileAdapter.getData(fileName)
    # liveStatus = True
    #
    # adk_conn = ADKconn()
    # datastreamId = adk_conn.createDataStream()
    # # datastreamId = "67r7y26cnncc8l"
    # adk_conn.postData(datastreamId, data, fileType, liveStatus)

    ########################################################################################





    ############### For creating datastream and adding historical data from a stream ################
    """
    The below code will create a datastream and post the historical data as a stream from
    the file to the datastream. You will have to give the fileName and pass it to the 
    get_data_stream() method of the file adapter.
    """

    # fileAdapter = FileAdapter()
    # fileName = "simple_sliding_mixed_multi_entity_source_moreData.csv"
    # data, fileType = fileAdapter.getDataStream(fileName)
    # liveStatus = True

    # adk_conn = ADKconn()
    # datastreamId = createDataStream()
    # adk_conn.postDataFromStream(datastreamId, data, fileType, liveStatus)

    ###############################################################################################




    ################### For live data input and output #########################
    """
    The below code will run both the functions of adding live input and getting
    live output simultaneously. You will have to enter the fileName from where
    you are getting the live data and pass it to the get_data_stream() method of
    the file adapter.
    
    NOTE:-
    1. Go on the Falkonry UI and build a model.
    2. After building a model click LIVE(OFF) button to turn on the LIVE INPUT
    """

    fileAdapter = FileAdapter()
    fileName = "source1.csv"
    data, fileType = fileAdapter.getDataStream(fileName)
    liveStatus = True

    adk_conn = ADKconn()### For live stream input
    p1 = Process(target=adk_conn.getLiveOutput, args=("j7kbqmmwywy78j", ))
    p1.start()
    p2 = Process(target=adk_conn.postDataFromStream, args=("kc88g9knrylyvn", data, fileType, liveStatus))
    p2.start()
    
    p1.join()
    p2.join()


    ###########################################################################




    ############## For creating datastream and adding data from a folder containing multiple files ################
    """
    The below code will create a datastream and post the historical data as a stream from
    the folder containing multiple files to the datastream. You will have to give the folder path and pass it to the 
    postMoreHistoricalDataFromStream() in the adk connector
    """

    # path = "../demo-data"
    # liveStatus = True
    #
    # adk_conn = ADKconn()
    # datastreamId = adk_conn.createDataStream()
    # adk_conn.postMoreDataFromStream(datastreamId, path, liveStatus)

    ##############################################################################################################

File: adk-examples/c-sharp/ConsoleApp2/ADKConnector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Threading;
using FalkonryClient;
using FalkonryClient.Helper.Models;
using Newtonsoft.Json;
using System.IO;
using log4net;

namespace ConsoleApp2
{
    class ADKConnnector
    {
        static String url = "https://example.falkonry.ai";
        static String token = "token";
        Falkonry falkonry = new Falkonry(url, token);
        DatastreamRequest ds = new DatastreamRequest();
        Datasource dataSource = new Datasource();
        Field field = new Field();
        Time time = new Time();
        Signal signal = new Signal();

        private int CheckStatus(System.String trackerId)
        {
            for (int i = 0; i < 12; i++)
            {
                Tracker tracker = falkonry.GetStatus(trackerId);
                if (tracker.Status.Equals("FAILED") || tracker.Status.Equals("ERROR"))
                {
                    throw new System.Exception(tracker.Message + "\n");
                }
                else if (tracker.Status.Equals("SUCCESS") || tracker.Status.Equals("COMPLETED"))
                {
                    Console.Write(tracker.Status + "\n");
                    return 1;
                }
                System.Threading.Thread.Sleep(5000);
            }
            return 0;
        }

        public String CreateDataStream()
        {
            String datastreamName = "C#_TEST_4";
            String timeZone = "GMT";
            String timeIdentifier = "time";
            String timeFormat = "millis";
            String PrecisionFormat = "millis";
            String entityIdentifier = "entity";
            //String signalIdentifier = "signal";     //for narrow data format
            //String valueIdentifier = "value";       //for narrow data format
            //String batchIdentifier = "batch_id";       //for batch-type data

            time.Format = timeFormat;
            time.Zone = timeZone;
            time.Identifier = timeIdentifier;
            ds.TimePrecision = PrecisionFormat;
            dataSource.Type = "PI";
            ds.DataSource = dataSource;
            ds.Name = datastreamName;
            ds.Field = field;
            ds.Field.Time = time;
            ds.Field.Signal = signal;
            ds.Field.EntityIdentifier = entityIdentifier;
            //ds.Field.Signal.SignalIdentifier = signalIdentifier;        //for narrow data format
            //ds.Field.Signal.ValueIdentifier = valueIdentifier;          //for narrow data format
            //ds.Field.BatchIdentifier = batchIdentifier;                 //for batch-type data
            Datastream datastream = falkonry.CreateDatastream(ds);
            Console.Write("New Datastream with id {0} created. \n", datastream.Id);
            return datastream.Id;
        }

        public void PostData(String datastreamId, String stream, String fileType, Boolean liveStatus=false)
        {
            if (fileType == null || datastreamId == null || stream == null)
            {
                throw new ArgumentNullException(nameof(fileType));
            }
            Datastream datastream = falkonry.GetDatastream(datastreamId);
            var options = new SortedDictionary<string, string>();
            var inputStatus = new InputStatus();
            options.Add("streaming", liveStatus.ToString());
            options.Add("hasMoreData", "false");
            options.Add("timeIdentifier", datastream.Field.Time.Identifier.ToString());
            options.Add("timeZone", datastream.Field.Time.Zone.ToString());
            options.Add("timeFormat", datastream.Field.Time.Format.ToString());
            options.Add("entityIdentifier", datastream.Field.EntityIdentifier.ToString());
            //options.Add("signalIdentifier", datastream.Field.Signal.SignalIdentifier.ToString());     // for narrow data format
            //options.Add("valueIdentifier", datastream.Field.Signal.ValueIdentifier.ToString());       // for narrow data format
            //options.Add("batchIdentifier", datastream.Field.BatchIdentifier.ToString());              // for batch-type data
            options.Add("fileFormat", fileType);
            if (!liveStatus)
            {
                int i;
                for (i = 0; i < 3; i++)
                {
                    try
                    {
                        inputStatus = falkonry.AddInput(datastreamId, stream, options);
                        if (CheckStatus(inputStatus.Id) == 1)
                        {
                            break;
                        }
                    }
                    catch (FalkonryClient.Service.FalkonryException e)
                    {
                        log.Error(e.GetBaseException() + "\n");
                        Console.ReadKey();
                        Environment.Exit(1);
                    }
                    catch (Exception ex)
                    {
                       log.WarnFormat(ex.Message + "\n" + "Retry Attempt: {0}", i + 1);
                    }
                }
                if (i == 3)
                {
                    log.Error("Cannot add data to the datastream!");
                    Console.ReadKey();
                    Environment.Exit(1);
                }
            }
            else
            {
                try
                {
                    inputStatus = falkonry.AddInput(datastreamId, stream, options);
                }
                catch (FalkonryClient.Service.FalkonryException e)
                {
                    log.Error(e.GetBaseException() + "\n");
                }
            }
        }

        public void PostDataFromStream(String datastreamId, byte[] stream, String fileType, Boolean liveStatus = false)
        {
            if (fileType == null || datastreamId == null || stream == null)
            {
                throw new ArgumentNullException(nameof(fileType));
            }
            Datastream datastream = falkonry.GetDatastream(datastreamId);
            var inputStatus = new InputStatus();
            var options = new SortedDictionary<string, string>();
            options.Add("streaming", liveStatus.ToString());
            options.Add("hasMoreData", "false");
            options.Add("timeIdentifier", datastream.Field.Time.Identifier.ToString());
            options.Add("timeZone", datastream.Field.Time.Zone.ToString());
            options.Add("timeFormat", datastream.Field.Time.Format.ToString());
            options.Add("entityIdentifier", datastream.Field.EntityIdentifier.ToString());
            //options.Add("signalIdentifier", datastream.Field.Signal.SignalIdentifier.ToString());     // for narrow data format
            //options.Add("valueIdentifier", datastream.Field.Signal.ValueIdentifier.ToString());       // for narrow data format
            //options.Add("batchIdentifier", datastream.Field.BatchIdentifier.ToString());              // for batch-type data
            options.Add("fileFormat", fileType);
            if (!liveStatus)
            {
                int i;
                for (i = 0; i < 3; i++)
                {
                    try
                    {
                        inputStatus = falkonry.AddInputStream(datastreamId, stream, options);
                        if (CheckStatus(inputStatus.Id) == 1)
                        {
                            break;
                        }
                    }
                    catch (FalkonryClient.Service.FalkonryException e)
                    {
                        log.Error(e.GetBaseException() + "\n");
                        Console.ReadKey();
                        Environment.Exit(1);
                    }
                    catch (Exception ex)
                    {
                        log.WarnFormat(ex.Message + "\n" + "Retry Attempt: {0}", i + 1);
                    }
                }
                if(i == 3)
                {
                    log.Error("Cannot add data to the datastream!");
                    Console.ReadKey();
                    Environment.Exit(1);
                }
            }
            else
            {
                try
                {
                    inputStatus = falkonry.AddInputStream(datastreamId, stream, options);
                }
                catch (FalkonryClient.Service.FalkonryException e)
                {
                    Console.Write(e.GetBaseException() + "\n");
                }
            }
        }


        public void PostMoreDataFromStream(String datastreamId, String folderPath, Boolean liveStatus = false)
        {
            if (datastreamId == null || folderPath == null)
            {
                throw new ArgumentNullException(nameof(datastreamId));
            }
            FileAdapter fileAdapter = new FileAdapter();
            Datastream datastream = falkonry.GetDatastream(datastreamId);
            var options = new SortedDictionary<string, string>();
            options.Add("streaming", liveStatus.ToString());
            options.Add("hasMoreData", "true");
            options.Add("timeIdentifier", datastream.Field.Time.Identifier.ToString());
            options.Add("timeZone", datastream.Field.Time.Zone.ToString());
            options.Add("timeFormat", datastream.Field.Time.Format.ToString());
            options.Add("entityIdentifier", datastream.Field.EntityIdentifier.ToString());
            //options.Add("signalIdentifier", datastream.Field.Signal.SignalIdentifier.ToString());     // for narrow data format
            //options.Add("valueIdentifier", datastream.Field.Signal.ValueIdentifier.ToString());       // for narrow data format
            //options.Add("batchIdentifier", datastream.Field.BatchIdentifier.ToString());              // for batch-type data
            options.Add("fileFormat", "csv");
            var inputStatus = new InputStatus();
            String[] Files = Directory.GetFiles(folderPath + "\\", "*.csv");
            var fileCount = Files.Length;
            foreach(String file in Files)
            {
                byte[] stream = fileAdapter.GetStream(file);
                if(fileCount == 1)
                {
                    options["hasMoreData"] = "false";
                }
                fileCount--;
                if (!liveStatus)
                {
                    int i;
                    for (i = 0; i < 3; i++)
                    {
                        try
                        {
                            inputStatus = falkonry.AddInputStream(datastreamId, stream, options);
                            if (CheckStatus(inputStatus.Id) == 1)
                            {
                                break;
                            }
                        }
                        catch (FalkonryClient.Service.FalkonryException e)
                        {
                            log.Error(e.GetBaseException() + "\n");
                            Console.ReadKey();
                            Environment.Exit(1);
                        }
                        catch (Exception ex)
                        {
                            log.WarnFormat(ex.Message + "\n" + "Retry Attempt: {0}", i + 1);
                        }
                    }
                    if (i == 3)
                    {
                        log.Error("Cannot add data to the datastream!");
                        Console.ReadKey();
                        Environment.Exit(1);
                    }
                }
                else
                {
                    try
                    {
                        inputStatus = falkonry.AddInputStream(datastreamId, stream, options);
                    }
                    catch (FalkonryClient.Service.FalkonryException e)
                    {
                        log.Error(e.GetBaseException() + "\n");
                    }
                }
            }
        }

        public void GetLiveOutput(String assessmentId)
        {
            void EventSource_Message(object sender, EventSource.ServerSentEventArgs e)
            {
                try
                {
                    var falkonryEvent = JsonConvert.DeserializeObject<FalkonryEvent>(e.Data);
                    log.Info(falkonryEvent.ToString() + "\n");
                }
                catch (System.Exception exception)
                {
                    log.Error(exception.Message);
                }

            }

            //On any error while getting live streaming output, EventSource_Error will be triggered
            void EventSource_Error(object sender, EventSource.ServerSentErrorEventArgs e)
            {
                Console.Write(e.Exception.Message + "\n");
                Console.Write(e.Exception.StackTrace + "\n");
            }

            EventSource eventSource = falkonry.GetOutput(assessmentId, null, null);
            eventSource.Message += EventSource_Message;
            eventSource.Error += EventSource_Error;

            log.Info("Press any key to stop listening.\n");
            Console.ReadKey();
            //eventSource.Dispose();
        }

        private static readonly ILog log = LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

        public static void Main(String[] args)
        {
            /*
                 USAGE:-
                 File adapter will be used to get the appropriate data from the different files and to provide it
                 in the format in which we can send to the ADK methods.

                *************** PLEASE NOTE ALL DATA FILES TO BE KEPT INSIDE THE 'ConsoleApp2/bin/debug/' FOLDER for relative path purposes.***************
             */

            /*
                #################### For creating datastream and adding data ################
                The below code will create a datastream and post the historical data as a string to the
                datastream. You will have to give the data as a string or give the fileName and pass it
                to the getData() method of the file adapter.
            */

            //FileAdapter fileAdapter = new FileAdapter();
            //ADKConnnector aDK = new ADKConnnector();

            //var datastreamId = aDK.CreateDataStream();

            //String fileName = "fileName";
            //var stream = fileAdapter.GetData(fileName);
            //String fileType = "csv";
            //aDK.PostData(datastreamId, stream, fileType, liveStatus=false);

            //  ###############################################################################

            /*
                #################### For creating datastream and adding data from a stream  ################
                The below code will create a datastream and post the historical data as a stream from
                the file to the datastream. You will have to give the fileName and pass it to the
                getDataStream() method of the file adapter.
            */

            //FileAdapter fileAdapter = new FileAdapter();
            //ADKConnnector aDK = new ADKConnnector();

            //var datastreamId = aDK.CreateDataStream();

            //String fileName = "fileName";
            //var stream = fileAdapter.GetStream(fileName);
            //String fileType = "csv";
            //aDK.PostDataFromStream(datastreamId, stream, fileType, liveStatus = false);

            //  #############################################################################################

            /*
                #################### For live data input and output  ################
                The below code will create a datastream and post the historical data as a stream from
                the file to the datastream. You will have to give the fileName and pass it to the
                getDataStream() method of the file adapter.
                NOTE:-
                 1. Go on the Falkonry UI and build a model.
                 2. After building a model, run the code.
                 While using this method you can choose between postData(), postDataFromStream() and postMoreDataFromStream()
                 and set live parameter as true.
            */

            //FileAdapter fileAdapter = new FileAdapter();
            //ADKConnnector aDK = new ADKConnnector();

            //String datastreamId = "datastreamId";
            //String assessmentId = "assessmentId";

            //aDK.TurnOnLiveMonitoring(datastreamId);

            //String fileName = "fileName";
            //var stream = fileAdapter.GetStream(fileName);
            //String fileType = "csv";
            //aDK.PostDataFromStream(datastreamId, stream, fileType, liveStatus = true);

            //Parallel.Invoke(() =>
            //{
            //    aDK.PostDataFromStream(datastreamId, stream, fileType, liveStatus = true);
            //},
            //() =>
            //{
            //    aDK.GetLiveOutput(assessmentId);
            //});

            //aDK.TurnOffLiveMonitoring(datastreamId);

            //  #########################################################################

            /*
                #################### Example For creating datastream and adding data from a folder containing multiple files   ################
  
                The below code will create a datastream and post the historical data as a stream from
                the folder containing multiple files to the datastream. You will have to give the folder path and pass it to the
                addMoreHistoricalDataFromStream() in the adk connector
            */
            //ADKConnnector aDK = new ADKConnnector();

            //String folderPath = "folderPath";
            //String datastreamId = "datastreamId";
            //aDK.PostMoreDataFromStream(datastreamId, folderPath, liveStatus = false);

            //  ########################################################################################################

        }
    }
}
